packages/core/src/components/context-menu/contextMenuTarget.tsx

```diff
@@ -17,45 +18,44 @@ export interface IContextMenuTarget extends React.Component<any, any> {
     onContextMenuClose?(): void;
 }
 
-export function ContextMenuTarget<T extends { prototype: IContextMenuTarget }>(constructor: T) {
-    const { render, renderContextMenu, onContextMenuClose } = constructor.prototype;
-
-    if (!isFunction(renderContextMenu)) {
+export function ContextMenuTarget<T extends IConstructor<IContextMenuTarget>>(WrappedComponent: T) {
+    if (!isFunction(WrappedComponent.prototype.renderContextMenu)) {
         console.warn(CONTEXTMENU_WARN_DECORATOR_NO_METHOD);
     }
 
-    // patching classes like this requires preserving function context
-    // tslint:disable-next-line only-arrow-functions
-    constructor.prototype.render = function(this: IContextMenuTarget) {
-        /* tslint:disable:no-invalid-this */
-        const element = render.call(this) as JSX.Element;
+    return class ContextMenuTargetClass extends WrappedComponent {
+        public static displayName = `ContextMenuTarget(${getDisplayName(WrappedComponent)})`;
 
-        if (element == null) {
-            // always return `element` in case caller is distinguishing between `null` and `undefined`
-            return element;
-        }
+        public render() {
+            // TODO handle being applied on a Component that doesn't return an actual Element
```

Comment:
this code handles `null | undefined | JSX.Element`. I think we should throw an error if `element` is not one of those types (namely, `string | number | Fragment`)

`"@ContextMenuTarget-decorated components must return a single JSX.Element or an empty render."`

Reply:
is it possible to support `React.ReactNode` return types? that would include `React.ReactPortal`.

Reply:
string/number/portal/fragment don't have real DOM elements so there's no place to put the onKeyDown/onKeyUp handlers (unless we add our own wrapper `<div>` but I'm pretty opposed to this DOM monkeypatching)

Reply:
ok fair enough